LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/LoRaDeviceClientExtensions.cs

```diff
@@ -0,0 +1,125 @@
+        }
+
+        private sealed class ResilientClient : ILoRaDeviceClient, IIdentityProvider<ILoRaDeviceClient>
+        {
+            private const int MaxAttempts = 3;
+
+            private readonly ILoRaDeviceClient client;
+            private readonly ILogger? logger;
+
+            public ResilientClient(ILoRaDeviceClient client, ILogger<ResilientClient>? logger)
+            {
+                this.client = client;
+                this.logger = logger;
+            }
+
+            public bool EnsureConnected() => this.client.EnsureConnected();
+            public Task DisconnectAsync(CancellationToken cancellationToken) => this.client.DisconnectAsync(cancellationToken);
+            ValueTask IAsyncDisposable.DisposeAsync() => this.client.DisposeAsync();
+
+            private sealed class Void
+            {
+                public static readonly Void Value = new();
+                private Void() { }
+            }
+
+            private Task<TResult> InvokeAsync<TResult>(Operation<Void, Void, TResult> operation,
+                                                       CancellationToken cancellationToken,
+                                                       [CallerMemberName] string? operationName = null) =>
+                InvokeAsync(operation, Void.Value, Void.Value, cancellationToken, operationName);
+
+            private Task<TResult> InvokeAsync<T, TResult>(Operation<T, Void, TResult> operation, T arg,
+                                                          CancellationToken cancellationToken,
+                                                          [CallerMemberName] string? operationName = null) =>
+                InvokeAsync(operation, arg, Void.Value, cancellationToken, operationName);
+
+            private async Task<TResult> InvokeAsync<T1, T2, TResult>(Operation<T1, T2, TResult> operation, T1 arg1, T2 arg2,
+                                                                     CancellationToken cancellationToken,
+                                                                     [CallerMemberName] string? operationName = null)
+            {
+                for (var attempt = 1; ; attempt++)
+                {
+                    try
+                    {
+                        _ = this.client.EnsureConnected();
+                        return await operation(this.client, arg1, arg2, cancellationToken);
+                    }
+                    catch (Exception ex)
+                        when ((ex is ObjectDisposedException
+                                  || (ex is InvalidOperationException ioe
+                                      && ioe.Message.StartsWith("This operation is only allowed using a successfully authenticated context.", StringComparison.OrdinalIgnoreCase)))
+                              && ExceptionFilterUtility.True(() =>
+                                     this.logger?.LogWarning(ex, @"Device client operation ""{Operation}"" (attempt {Attempt}/"
```

Comment:
Since we are handling the exception, let's not log a warning. As long as we can handle it, we are ok and if we can't (due to the retries) we will log it up in the stack and we know we ran out of retries. 

I was contemplating if a metric of the retries would make sense but I think it does not add much value.

Reply:
Addressed with f845f60e700f0eef570879b71c3f951e0f62308e.